### PR TITLE
Fix ledger-tool program command for non-runnable subcommands

### PR DESCRIPTION
**Pull request by dmakarov**
#### Problem

The cfg and disassemble sub-sub-commands don't need to load a ledger database for running, and neither do they need an input file specifying instruction accounts etc.

#### Summary of Changes

This change creates a simplified invoke_context for processing the static actions of ledger-tool program sub-command.

### CODE REVIEW

**Comment by dmakarov** [comment]
done with refactoring. should be good to review.

File: ledger-tool/src/program.rs [hunk 1]
```diff
@@ -324,16 +324,117 @@ fn output_trace(
     }
 }
 
+macro_rules! load_program {
```
Comment (steviez) [inline ledger-tool/src/program.rs:327]: There is probably a reason that isn't obvious to me from just scanning the PR without pulling the code down locally - why go with a macro over here over a function?

Comment (dmakarov) [reply ledger-tool/src/program.rs:327]: If load_program is a function then invoke_context is borrowed mut more than once.

Comment (steviez) [reply ledger-tool/src/program.rs:327]: Hmm, I'm not sure I see the problem. I pulled your branch down locally, and defined the following function:
```rust
fn load_program<'a>(
    filename: &Path,
    program_id: Pubkey,
    invoke_context: &InvokeContext<'a>,
) -> Executable<RequisiteVerifier, InvokeContext<'a>> {
    let mut file = File::open(filename).unwrap();
    let mut magic = [0u8; 4];
    file.read_exact(&mut magic).unwrap();
    file.rewind().unwrap();
    let is_elf = magic == [0x7f, 0x45, 0x4c, 0x46];
    let mut contents = Vec::new();
    file.read_to_end(&mut contents).unwrap();
    let slot = Slot::default();
    let reject_deployment_of_broken_elfs = false;
    let debugging_features = true;
    let log_collector = invoke_context.get_log_collector();
    let loader_key = bpf_loader_upgradeable::id();
    let mut load_program_metrics = LoadProgramMetrics {
        program_id: program_id.to_string(),
        ..LoadProgramMetrics::default()
    };
    let account_size = contents.len();
    let mut verified_executable = if is_elf {
        let result = load_program_from_bytes(
            &invoke_context.feature_set,
            invoke_context.get_compute_budget(),
            log_collector,
            &mut load_program_metrics,
            &contents,
            &loader_key,
            account_size,
            slot,
            reject_deployment_of_broken_elfs,
            debugging_features,
        );
        match result {
            Ok(loaded_program) => match loaded_program.program {
                LoadedProgramType::LegacyV1(program) => Ok(unsafe { std::mem::transmute(program) }),
                _ => unreachable!(),
            },
            Err(err) => Err(format!("Loading executable failed: {err:?}")),
        }
    } else {
        let loader = create_loader(
            &invoke_context.feature_set,
            invoke_context.get_compute_budget(),
            true,
            true,
        )
        .unwrap();
        let executable =
            assemble::<InvokeContext>(std::str::from_utf8(contents.as_slice()).unwrap(), loader)
                .unwrap();
        Executable::<RequisiteVerifier, InvokeContext>::verified(executable)
            .map_err(|err| format!("Assembling executable failed: {err:?}"))
    }
    .unwrap();
    #[cfg(all(not(target_os = "windows"), target_arch = "x86_64"))]
    verified_executable.jit_compile().unwrap();
    verified_executable
}
```
which I then use like this:
```rust
    let verified_executable = load_program(
        Path::new(program),
        program_id,
        &invoke_context,
    );
    let mut analysis = LazyAnalysis::new(&verified_executable);

    match action {
        Action::Cfg => {
            let mut file = File::create("cfg.dot").unwrap();
            analysis
                .analyze()
                .visualize_graphically(&mut file, None)
                .unwrap();
        }
        Action::Dis => {
            let stdout = std::io::stdout();
            analysis.analyze().disassemble(&mut stdout.lock()).unwrap();
        }
    };
```
and
```rust
    let verified_executable = load_program(
        Path::new(program),
        program_id,
        &invoke_context,
    );
    let mut analysis = LazyAnalysis::new(&verified_executable);

    create_vm!(
        vm,
        &verified_executable,
        regions,
        account_lengths,
        &mut invoke_context,
    );
    ...
```

This compiles fine for me

Comment (dmakarov) [reply ledger-tool/src/program.rs]: Yes, this seems to work. Not sure what I did wrong when I got the borrow mut error.  I updated the macro to be a function.

Comment (steviez) [reply ledger-tool/src/program.rs]: I was initially passing in the InvokeContext by value which gave me:
```
error[E0382]: borrow of moved value: `invoke_context`
```
Not sure if that is what you hit but all good regardless. Additionally, I pulled this block out:
```
let mut analysis = LazyAnalysis::new(&verified_executable);
```
Doing so avoided a lifetime issue that probably could have been resolved; however, I think having the analysis item get created separately is a good separation. Ie, `load_program()` just loads the program

**Pull request merged**